src/snowflake/cli/plugins/nativeapp/manager.py

```diff
@@ -578,12 +592,75 @@ def deploy(
             self._apply_package_scripts()
 
             # 3. Upload files from deploy root local folder to the above stage
+            stage_fqn = stage_fqn or self.stage_fqn
             diff = self.sync_deploy_root_with_stage(
                 bundle_map=bundle_map,
                 role=self.package_role,
                 prune=prune,
                 recursive=recursive,
+                stage_fqn=stage_fqn,
                 local_paths_to_sync=local_paths_to_sync,
             )
 
+        if validate:
+            self.validate(use_scratch_stage=False)
+
         return diff
+
+    def validate(self, use_scratch_stage: bool = False):
+        """Validates Native App setup script SQL."""
+        cc.step(f"Validating Snowflake Native App setup script.")
+        validation_result = self.get_validation_result(use_scratch_stage)
+
+        # First print warnings, regardless of the outcome of validation
+        for warning in validation_result.get("warnings", []):
+            cc.warning(_validation_item_to_str(warning))
+
+        # Then raise an exception if validation failed
+        if validation_result["status"] == "FAIL":
+            messages = [
+                _validation_item_to_str(error)
+                for error in validation_result.get("errors", [])
+            ]
+            raise SetupScriptFailedValidation(messages)
+
+    def get_validation_result(self, use_scratch_stage: bool):
+        """Call system$validate_native_app_setup() to validate deployed Native App setup script."""
+        stage_fqn = self.stage_fqn
+        if use_scratch_stage:
+            stage_fqn = self.scratch_stage_fqn
+            bundle_map = self.build_bundle()
+            self.deploy(
+                bundle_map=bundle_map,
+                prune=True,
+                recursive=True,
+                stage_fqn=stage_fqn,
+                validate=False,
+            )
+        prefixed_stage_fqn = StageManager.get_standard_stage_prefix(stage_fqn)
+        try:
+            cursor = self._execute_query(
+                f"call system$validate_native_app_setup('{prefixed_stage_fqn}')"
+            )
+        except ProgrammingError as err:
+            if err.errno == 2003 and ERROR_MESSAGE_2003 in err.msg:
+                raise ApplicationPackageDoesNotExistError(self.package_name)
+            generic_sql_error_handler(err)
+        else:
+            if not cursor.rowcount:
+                raise SnowflakeSQLExecutionError()
+            return json.loads(cursor.fetchone()[0])
+        finally:
+            if use_scratch_stage:
+                cc.step(f"Dropping stage {self.scratch_stage_fqn}.")
+                with self.use_role(self.package_role):
+                    self._execute_query(
+                        f"drop stage if exists {self.scratch_stage_fqn}"
+                    )
+
+
+def _validation_item_to_str(item: dict[str, str | int]):
```

Comment:
For the user's convenience, IMHO we should include the line number and filename too. IIRC these are both present in the blob we get back from the validation API.

Reply:
They're already included in the error/warning message, and where they aren't it's because they aren't known (see the "App has validation errors" screenshot in the description)

Reply:
oh right, I remember now, thanks.